splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+            }
+        if command == b'PKSIGN':
+            return {
+                b'PINENTRY_LAUNCHED': self.inquire_PINENTRY_LAUNCHED,
+            }
+        return {}
+
+    # was handle_server_response in ruby version
+    async def send_command(self, command: bytes, args: Optional[bytes]):
+        assert self.server.state == ServerState.client_command
+        self.server.state = ServerState.agent_response
+        try:
+            self.expected_inquires = self.get_inquires_for_command(command)
+            loop = asyncio.get_event_loop()
+            self.agent_responded = loop.create_future()
+            if args:
+                cmd_with_args = command + b' ' + args + b'\n'
+            else:
+                cmd_with_args = command + b'\n'
+            self.agent_write(cmd_with_args)
+            await self.agent_responded
+        finally:
+            self.server.state = ServerState.client_command
+
+    def agent_write(self, data):
+        self.server.log_io('A <<<', data)
+        self.agent_transport.write(data)
+
+    def data_received(self, untrusted_data: bytes) -> None:
+        # pylint: disable=arguments-differ
+        self.server.log_io('A >>>', untrusted_data)
+        if self.server.state != ServerState.agent_response:
+            self.server.abort('message from gpg-agent at unexpected time')
+            return
+
+        self.untrusted_agent_response_buffer.extend(untrusted_data)
+        while b'\n' in self.untrusted_agent_response_buffer:
+            untrusted_line, untrusted_rest = \
+                self.untrusted_agent_response_buffer.split(b'\n', 1)
+            self.untrusted_agent_response_buffer = untrusted_rest
+            # convert from bytesarray to bytes to be hashable
+            untrusted_line = bytes(untrusted_line)
+            try:
+                self.handle_agent_response(untrusted_line=untrusted_line)
+            except Exception as e:  # pylint: disable=broad-except
+                # propagate exception back to self.send_command
+                self.agent_responded.set_exception(e)
+                break
+
+    def handle_agent_response(self, *, untrusted_line):
+        try:
+            untrusted_res, untrusted_args = untrusted_line.split(b' ', 1)
+        except ValueError:
+            untrusted_res = untrusted_line
+            untrusted_args = None
+        if untrusted_res in (b'D', b'S'):
+            # passthrough to the client
+            self.server.client_write(untrusted_line + b'\n')
+        elif untrusted_res in (b'OK', b'ERR'):
+            # passthrough to the client and signal command complete
+            self.server.client_write(untrusted_line + b'\n')
+            try:
+                self.agent_responded.set_result(None)
+            except asyncio.InvalidStateError:
+                if untrusted_res == b'ERR':
+                    # agent may report multiple errors
+                    pass
```

Comment:
Hmm, the old code didn't allow this. Why do we here?

Reply:
I think in testing I've got such situation and the second error message was more meaningful. But I don't remember details, so I'll drop it.